### PR TITLE
feat: load number of pages from disk in merge strategy

### PR DESCRIPTION
**Pull request by pakhomov-dfinity**
In order to create merge strategy, we need to know for each `PageMap` its size on disk and in memory. We used to propagate the in memory size from DSM thread. It is complicated code wise and makes the lazy `PageMaps` loading harder.
This change loads pagemaps in the `TipThread` in parallel to get the in-memory size.

### CODE REVIEW

**Comment by schneiderstefan** [comment]
Hmm. It seems excessively wasteful to open and mmap potentially hundreds of thousands of overlay files, and then verify them, just to get a number that we already have and could collect in a few milliseconds.

I think it would be much better to just add a function `PageMapType::size` that returns the size according to execution and call that in `pagemaptypes_with_num_pages` instead. That would not interfere with lazy loading at all. Furthermore, we only need these sizes to ensure a limit on the storage overhead. But for that, what we really care about is the overhead vs the state size that execution enforces. So it's the size according to execution that we really care about anyway, and not the largest page index + 1.

It's true that the work happens in the tip thread, where it is less critical, but we still spend it on unnecessary work that has to run in parallel to execution.

**Review by basvandijk — APPROVED** [review]

File: rs/replicated_state/src/page_map/storage.rs [hunk 1]
```diff
@@ -884,6 +886,70 @@ impl dyn StorageLayout + '_ {
         }
         Ok(result)
     }
+
+    // Read the number of memory pages from overlay.
+    // Basically it's the index of the last page, which we read based on the offset from the end of
+    // the file plus some error handling.
+    fn overlay_memory_pages(overlay: &Path) -> StorageResult<usize> {
```
Comment (schneiderstefan) [inline rs/replicated_state/src/page_map/storage.rs:893]: The naming is a bit ambiguous (number of non-zero pages vs number of logical pages), but I can't really think of a strictly better name. We called it `num_logical_pages` somewhere else.

File: rs/replicated_state/src/page_map/storage.rs [hunk 2]
```diff
@@ -884,6 +886,70 @@ impl dyn StorageLayout + '_ {
...
+            -((VERSION_NUM_BYTES + SIZE_NUM_BYTES + PAGE_INDEX_RANGE_NUM_BYTES) as i64),
+        ))
+        .map_err(|err| to_storage_err(err))?;
+        file.read_exact(&mut last_page_index_range_buf.as_flattened_mut())
+            .map_err(|err| to_storage_err(err))?;
+        let last_page_index_range = PageIndexRange::from(&last_page_index_range_buf);
+        Ok(last_page_index_range.end_page.get() as usize)
+    }
+
+    pub fn memory_pages(&self) -> StorageResult<usize> {
```
Comment (schneiderstefan) [inline rs/replicated_state/src/page_map/storage.rs:933]: Doc comment would be good.

File: rs/state_manager/src/tip.rs [hunk 3]
```diff
@@ -584,32 +584,63 @@ struct StorageInfo {
     mem_size: u64,
 }
 
+impl StorageInfo {
+    fn add(&self, rhs: &StorageInfo) -> StorageInfo {
```
Comment (schneiderstefan) [inline rs/state_manager/src/tip.rs:588]: You could also consider implementing the trait `std::ops::Add`.

**Review by schneiderstefan — APPROVED** [review]
Just some minor comments.

**Pull request merged**